orttraining/orttraining/training_ops/cuda/communication/send.cc

```diff
@@ -28,145 +28,192 @@ ONNX_OPERATOR_KERNEL_EX(
-
-  // Keep the sync copy in the previous design
-  // TODO they can be moved to async call after global stream becoming accessible
-  for (int i = 0; i < tensor_num; ++i) {
-    const Tensor* x_tensor = ctx->Input<Tensor>(i + 2);
-    ORT_ENFORCE(cudaMemcpy(buffer.get() + tensor_offsets_in_bytes[i], x_tensor->DataRaw(),
-                           tensor_sizes_in_bytes[i], cudaMemcpyDeviceToHost) == cudaSuccess);
-  }
-
-  // Prepare MPI communication info
-  int tensor_num_in_bytes = tensor_num * static_cast<int>(sizeof(size_t));
-  ORT_ENFORCE(tensor_num_in_bytes < INT_MAX,
+void Send::SendShapeInfo(
+    const int dst,
+    const int num_tensors,  // Number of sent tensors.
+    size_t aggregated_aligned_tensor_bytes,
+    std::vector<size_t> prefix_tensor_shape_sizes,
+    std::vector<int64_t> aggregated_tensor_shapes) const {
+  const int num_tensors_in_bytes = num_tensors * static_cast<int>(sizeof(size_t));
+  ORT_ENFORCE(num_tensors_in_bytes < INT_MAX,
               "Total tensor number larger than MPI size limit");
+
   CommInfo_t info_shape_sizes{prefix_tensor_shape_sizes.data(),
-                              tensor_num_in_bytes,
+                              num_tensors_in_bytes,
                               dst,
                               static_cast<int>(tag_)};
-
   ORT_ENFORCE(aggregated_aligned_tensor_bytes < INT_MAX,
               "Aggregated tensor size larger than MPI size limit");
+
   CommInfo_t info_aggregated_size{&aggregated_aligned_tensor_bytes,
                                   static_cast<int>(sizeof(size_t)),
                                   dst,
                                   static_cast<int>(tag_)};
 
   int total_tensor_dim_in_bytes = static_cast<int>(
-    aggregated_tensor_shapes.size()) * static_cast<int>(sizeof(int64_t));
+                                      aggregated_tensor_shapes.size()) *
+                                  static_cast<int>(sizeof(int64_t));
   ORT_ENFORCE(total_tensor_dim_in_bytes < INT_MAX,
               "Total dimensions of tensors larger than MPI size limit");
+
   CommInfo_t info_shapes{aggregated_tensor_shapes.data(),
                          total_tensor_dim_in_bytes,
                          dst,
                          static_cast<int>(tag_)};
 
-  CommInfo_t info_data{buffer.get(),
-                       static_cast<int>(aggregated_aligned_tensor_bytes),
-                       dst,
-                       static_cast<int>(tag_)};
-
+  int mpi_code = 0;
 
   // Directly use CPU to wait MPI_Send. We cannot use GPU callback because
   // MPI_Send may block the entire GPU until it returns.
-  MPI_CHECK(MPI_Send(
-    info_shape_sizes.buffer, info_shape_sizes.size, MPI_CHAR,
-    info_shape_sizes.rank, info_shape_sizes.tag, MPI_COMM_WORLD));
+  mpi_code = MPI_Send(
+      info_shape_sizes.buffer, info_shape_sizes.size, MPI_CHAR,
+      info_shape_sizes.rank, info_shape_sizes.tag, MPI_COMM_WORLD);
+  ORT_ENFORCE(mpi_code == MPI_SUCCESS, "MPI Send fails.");
```

Comment:
>ORT_ENFORCE(mpi_code == MPI_SUCCESS, "MPI Send fails."); [](start = 2, length = 56)

can use MPI_CHECK to check the return code.

Reply:
Thanks. I modified both send.cc and recv.cc.

---
In reply to: [459816734](https://github.com/microsoft/onnxruntime/pull/4547#discussion_r459816734) [](ancestors = 459816734)